wal/watcher_test.go

```diff
@@ -0,0 +1,504 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package wal
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/tsdb/labels"
+	"github.com/prometheus/tsdb/record"
+	"github.com/prometheus/tsdb/testutil"
+)
+
+var defaultRetryInterval = 100 * time.Millisecond
+var defaultRetries = 100
+
+// retry executes f() n times at each interval until it returns true.
+func retry(t *testing.T, interval time.Duration, n int, f func() bool) {
+	t.Helper()
+	ticker := time.NewTicker(interval)
+	for i := 0; i <= n; i++ {
+		if f() {
+			return
+		}
+		<-ticker.C
+	}
+	ticker.Stop()
+	t.Logf("function returned false")
+}
+
+type writeToMock struct {
+	samplesAppended      int
+	seriesLock           sync.Mutex
+	seriesSegmentIndexes map[uint64]int
+}
+
+func (wtm *writeToMock) Append(s []record.RefSample) bool {
+	wtm.samplesAppended += len(s)
+	return true
+}
+
+func (wtm *writeToMock) StoreSeries(series []record.RefSeries, index int) {
+	wtm.seriesLock.Lock()
+	defer wtm.seriesLock.Unlock()
+	for _, s := range series {
+		wtm.seriesSegmentIndexes[s.Ref] = index
+	}
+}
+
+func (wtm *writeToMock) SeriesReset(index int) {
+	// Check for series that are in segments older than the checkpoint
+	// that were not also present in the checkpoint.
+	wtm.seriesLock.Lock()
+	defer wtm.seriesLock.Unlock()
+	for k, v := range wtm.seriesSegmentIndexes {
+		if v < index {
+			delete(wtm.seriesSegmentIndexes, k)
+		}
+	}
+}
+
+func (wtm *writeToMock) checkNumLabels() int {
+	wtm.seriesLock.Lock()
+	defer wtm.seriesLock.Unlock()
+	return len(wtm.seriesSegmentIndexes)
+}
+
+func newWriteToMock() *writeToMock {
+	return &writeToMock{
+		seriesSegmentIndexes: make(map[uint64]int),
+	}
+}
+
+func TestTailSamples(t *testing.T) {
+	pageSize := 32 * 1024
+	const seriesCount = 10
+	const samplesCount = 250
+	now := time.Now()
+
+	dir, err := ioutil.TempDir("", "readCheckpoint")
+	testutil.Ok(t, err)
+	defer os.RemoveAll(dir)
+
+	wdir := path.Join(dir, "wal")
+	err = os.Mkdir(wdir, 0777)
+	testutil.Ok(t, err)
+
+	enc := record.Encoder{}
+	w, err := NewSize(nil, prometheus.DefaultRegisterer, wdir, 128*pageSize, false)
```

Comment:
Would you mind re-copying this file? I changed all the tests to run with compression enabled and disabled as part of the 2.11 release.

Reply:
yeah I need to recopy this and the watcher, quite a few small improvements have been made